README.md
[![Node.js CI][ci-img]][ci]
[![Leaflet 1.9.4][leaflet-img]][leaflet]

# Leaflet Plugins

A collection of accessibility and localisation/ translation plugins for [Leaflet][].

| Plugin               | Description                               | NPM     |
|----------------------|-------------------------------------------|---------|
|[Leaflet.a11y][]      | An accessibility and localization plugin  | [![NPM][npm-a-img]][npm-a] |
|[Leaflet.translate][] | Translations/ language packs for Leaflet core and plugins| [![NPM][npm-t-img]][npm-t] |
|[L.keyboard-help][]   | A control to display help for keyboard shortcuts  | [![NPM][npm-k-img]][npm-k] |

* See: [Changelog][]
* See: [Roadmap][]
* Demo: [nfreear.github.io/leaflet.plugins][demo]

## Aim

This plugin does _NOT_ replace [accessibility][] efforts in [core Leaflet][bugs]. It is a means to provide interim fixes, and potentially to test fixes.

The aim for this plugin is for it to become obsolete and to retire it, because sufficient accessibility fixes and enhancements have been made to core Leaflet. For more see [#14][].

## What is “A11y“?

“A11y“ is a numeronym – an abbreviation for accessibility. It is often pronounced “alley“ (see [#10][]).

## Leaflet.i18n plugin optional

If available, `Leaflet.a11y` uses the [Leaflet.i18n][] internationalisation plugin.

If the `Leaflet.i18n` plugin is not available, a shim is provided for the `L._()` translation function.

## What it does

The plugin currently does the following:

1. Set a `role` and `aria-label` on the Leaflet map container element (see [L-7193][]).
2. Manage keyboard focus when popups are opened and closed (see [L-8115][]).
3. Fix so non-interactive markers are correctly identified ([L-8116][])
4. Enable map controls (Zoom in, Zoom out, Close popup, ...) to be translated into a language that has been set with `setLocale` ([Leaflet.i18n][]). Mostly relevant for accessibility, as most text is hidden from visual users ([L-9092][]).

More to follow!

## Usage

Include or `import` core Leaflet and the plugins:

```html
<script src="path/to/leaflet.js"></script>
<script src="path/to/Leaflet.i18n.js"></script>
<script src="path/to/Leaflet.a11y.js"></script>
```

Then, simply set the `a11yPlugin` option when creating a map:

```js
var map = L.map('map', {
  a11yPlugin: true
  // ...
}).setView([51.505, -0.09], 13);

// ...
```

## Build and Test

```
npm install
npm run build
npm test
npm start
```

Feedback, contributions and pull requests are welcomed!

---
License: [MIT][].

[roadmap]: https://github.com/nfreear/leaflet.plugins/blob/main/ROADMAP.md
[changelog]: https://github.com/nfreear/leaflet.plugins/blob/main/CHANGELOG.md
[ci]: https://github.com/nfreear/leaflet.plugins/actions/workflows/node.js.yml
[ci-img]: https://github.com/nfreear/leaflet.plugins/actions/workflows/node.js.yml/badge.svg
[leaflet-img]: https://img.shields.io/badge/leaflet-1.9.4-green.svg?style=flat
[demo]: https://nfreear.github.io/leaflet.plugins/
[Leaflet]: https://leafletjs.com/
[accessibility]: https://leafletjs.com/examples/accessibility/
[Leaflet.i18n]: https://github.com/umap-project/Leaflet.i18n
[MIT x]: https://nfreear.mit-license.org/
[MIT]: https://github.com/nfreear/leaflet.plugins/blob/main/LICENSE
[#10]: https://github.com/nfreear/leaflet.plugins/issues/10
  "What is “a11y”? #10"
[#14]: https://github.com/nfreear/leaflet.plugins/issues/14
  "“I want to retire” #14"
[bugs]: https://github.com/Leaflet/Leaflet/labels/accessibility
[L-7193]: https://github.com/Leaflet/Leaflet/issues/7193
  "Make the leaflet-container a programmatically determinable element"
[L-8115]: https://github.com/Leaflet/Leaflet/issues/8115
  "Focus management between markers and popups"
[L-8116]: https://github.com/Leaflet/Leaflet/issues/8116
  "Discern interactive markers from non-interactive markers"
[L-9092]: https://github.com/Leaflet/Leaflet/issues/9092
  "feat: Add placeholder function for translation/localization/i18n to Leaflet"
[L-9087]: https://github.com/Leaflet/Leaflet/pull/9087
  "Add the 'Leaflet.a11y' to plugin list"
[Maps WCAG eval]: https://github.com/Malvoz/web-maps-wcag-evaluation
  "Web map tools WCAG 2.1 evaluation - A manual accessibility evaluation of popular web map tools."

[Leaflet.a11y]: https://github.com/nfreear/leaflet.a11y/tree/main/packages/leaflet.a11y
[Leaflet.translate]: https://github.com/nfreear/leaflet.a11y/tree/main/packages/Leaflet.translate
[L.keyboard-help]: https://github.com/nfreear/leaflet.a11y/tree/main/packages/L.keyboard-help

[npm-a]: https://www.npmjs.com/package/leaflet.a11y
[npm-a-img]: https://img.shields.io/npm/v/leaflet.a11y
[npm-t]: https://www.npmjs.com/package/leaflet.translate
[npm-t-img]: https://img.shields.io/npm/v/leaflet.translate
[npm-k]: https://www.npmjs.com/package/l.keyboard-help
[npm-k-img]: https://img.shields.io/npm/v/l.keyboard-help


CHANGELOG.md

# Release Notes

## Leaflet Plugins Version 0.6.0

* Release: 24-Oct-2023,
* First full mono-repo release - 3 plugins! ([#21][])
* Renamed repo to `leaflet.plugins` ([#22][])
* Fix browser compatibility (Firefox) - `Leaflet.a11y` ([#20][])
* Separate locale data and functionality into `Leaflet.translate` ([#18][])
* Document [roadmap][] (todos!)

## Leaflet.a11y Version 0.3.0

* Release: 26-Sep-2023,
* Do things the Leaflet way – adopt [`L.Map.addInitHook`][addInitHook] ([#8][])
* Allow for multiple maps on page! ([#8][])
* Convert plugin to ESM module / class, with a build script ([#8][])
* Use an "import map" ([#11][])
* Put locale import code in separate class ([#12][])

## Version 0.2.0

* Release: 22-Sep-2023.
* Fix: null lang and empty string role bugs ([#7][])
* Fix: only set focus when opened by a trigger element ([#7][])
* `Leaflet.i18n` plugin is optional ([#6][])
* README and documentation fixes ([#1][])

## Version 0.1.0

* Release: 10-Sep-2023.
* Initial release of [Leaflet.a11y][].
* Fixes 4+ accessibility issues in core Leaflet.js
* Set a `role` and `aria-label` on the Leaflet map container ([#2][], [L-7193][]).
* Manage keyboard focus when popups are opened and closed ([#3][], [L-8115][]).
* Fix so non-interactive markers are correctly identified ([#4][], [L-8116][])
* Localization - enable map controls (Zoom in, Close popup, ...) to be translated into a language that has been set with `setLocale` ([Leaflet.i18n][]) ([L-9092][]).

---

[readme]: https://github.com/nfreear/leaflet.plugins#readme
[roadmap]: https://github.com/nfreear/leaflet.plugins/blob/main/ROADMAP.md
[demo]: https://nfreear.github.io/leaflet.plugins
  "Example site using ‘Leaflet.a11y’ plugin"
[i18n plugin]: https://github.com/umap-project/Leaflet.i18n
[Leaflet.translate]: https://github.com/nfreear/Leaflet.translate "Legacy!"
[wcag eval]: https://github.com/Malvoz/web-maps-wcag-evaluation
  "Web map tools WCAG 2.1 evaluation, by @Malvoz (13-Sep-2021)"

[L bugs]: https://github.com/Leaflet/Leaflet/labels/accessibility
  "Leaflet issues and pull-requests tagged with ‘accessibility’"
[L-7193]: https://github.com/Leaflet/Leaflet/issues/7193
  "Make the leaflet-container a programmatically determinable element [a11y]"
[L-7527]: https://github.com/Leaflet/Leaflet/issues/7527
  "Markers that bind to popups should convey the ‘aria-expanded’ state of the popup [a11y]"
[L-7822]: https://github.com/Leaflet/Leaflet/issues/7822
  "Accessible vectors [a11y]"
[L-8251]: https://github.com/Leaflet/Leaflet/pull/8251
  "Make vectors focusable and labellable [PR][a11y]"
[L-7968]: https://github.com/Leaflet/Leaflet/issues/7968
  "Potentially set popups to role=’dialog’ [needs decision][a11y]"
[L-8115]: https://github.com/Leaflet/Leaflet/issues/8115
  "Focus management between markers and popups [a11y]"
[L-8116]: https://github.com/Leaflet/Leaflet/issues/8116
  "Discern interactive markers from non-interactive markers [a11y]"
[L-9087]: https://github.com/Leaflet/Leaflet/pull/9087
  "Add 'Leaflet.a11y' to plugin list [PR][a11y]"
[L-9092]: https://github.com/Leaflet/Leaflet/issues/9092
   "Add placeholder function for translation/localization/i18n to Leaflet [proposal][a11y]"
[i18n-6]: https://github.com/umap-project/Leaflet.i18n/issues/6
  "Proposal: Add placeholder function for translation/localization/i18n to core Leaflet"

[#1]: https://github.com/nfreear/leaflet.plugins/issues/1
  "Documentation, CI, Unit tests, linting (4 tasks) [doc]"
[#2]: https://github.com/nfreear/leaflet.a11y/issues/2
  "Map container needs an accessible name and role [a11y]"
[#3]: https://github.com/nfreear/leaflet.a11y/issues/3
  "Focus should be managed between markers and popups [a11y]"
[#4]: https://github.com/nfreear/leaflet.a11y/issues/4
  "Non-interactive markers should not have an interactive role [a11y]"
[#5]: https://github.com/nfreear/leaflet.plugins/issues/5
  "Deploy example to GitHub Pages and publish plugin (3 tasks) [doc]"
[#6]: https://github.com/nfreear/leaflet.plugins/issues/6
  "Be more explicit about optional use of ‘Leaflet.i18n’ plugin [doc]"
[#7]: https://github.com/nfreear/leaflet.plugins/issues/7
  "Fix null lang and empty string role bugs, etc. (4 tasks) [bug]"
[#8]: https://github.com/nfreear/leaflet.a11y/issues/8
  "Use 'addInitHook'; Convert plugin to an ES module, with build step #8 (4 tasks)"
[#9]: https://github.com/nfreear/leaflet.a11y/issues/9
  "Add support for 3rd-party plugins, for example, fullscreen"
[#10]: https://github.com/nfreear/leaflet.a11y/issues/10
  "What is “a11y”? #10 [doc]"
[#11]: https://github.com/nfreear/leaflet.a11y/issues/11
  "Use an import map"
[#12]: https://github.com/nfreear/leaflet.a11y/issues/12
  "Put locale import code in a separate class #12"
[#14]: https://github.com/nfreear/leaflet.a11y/issues/14
  "“I want to retire” #14 [doc]"
[#18]: https://github.com/nfreear/leaflet.plugins/issues/18
  "Separate locale data and functionality into separate plugin"
[#20]: https://github.com/nfreear/leaflet.plugins/issues/20
  "Fix and document browser compatibility - Firefox etc."
[#21]: https://github.com/nfreear/leaflet.plugins/issues/21 "Mono-repo ..."
[addInitHook]: https://leafletjs.com/reference.html#class-constructor-hooks
  "Leaflet API – Class – Constructor hooks"

[End]: //


Leaflet.a11y.js
// Following https://github.com/Leaflet/Leaflet/blob/master/PLUGIN-GUIDE.md

(function (factory, window) {
  // define an AMD module that relies on 'leaflet'
  if (typeof define === 'function' && define.amd) { // eslint-disable-line no-undef
    define(['leaflet'], factory); // eslint-disable-line no-undef

    // define a Common JS module that relies on 'leaflet'
  } else if (typeof exports === 'object') {
    module.exports = factory(require('leaflet'));
  }

  // attach your plugin to the global 'L' variable
  if (typeof window !== 'undefined' && window.L) {
    window.L.a11y = factory(L); // eslint-disable-line no-undef
  }
}(function (L) {
  const MyAccessibilityPlugin = { onLoad, initialize, _qt };

  // implement your plugin
  let mapElem;
  let _initialized = false;

  // Dummy/placeholder "translation" function.
  L._ = L._ || ((str) => str);

  function onLoad (ev) {
    console.assert(ev && ev.type && ev.target && ev.target.on);
    let layer = 0;
    const MAP = ev.target;
    MAP.on('layeradd', ev => {
      // Initialize after the tile layer is added.
      if (layer === 1) {
        initialize(MAP);
      }
      layer++;
    });

    _fixMarkers(MAP);
  }

  function initialize (MAP) {
    console.assert(MAP && MAP._leaflet_id && MAP.getContainer);
    if (_initialized) return;
    _initialized = true;

    mapElem = MAP.getContainer(); // Was: MAP._container;

    mapElem.lang = L.locale || '';

    _localizeControls();
    _localizePopups(MAP); // NOT 'mapElem'!

    _fixMapContainer();
    _managePopupFocus(MAP);

    console.debug('a11y.initialize:', mapElem, [mapElem], MAP.getPanes(), MAP);
  }

  /** Translate default controls - Zoom in/out, Attribution.
   */
  function _localizeControls () {
    _qt('.leaflet-control-zoom-in', 'ariaLabel', L._('Zoom in')); // src/control/Control.Zoom.js#L30
    _qt('.leaflet-control-zoom-in', 'title', L._('Zoom in'));

    _qt('.leaflet-control-zoom-out', 'ariaLabel', L._('Zoom out'));
    _qt('.leaflet-control-zoom-out', 'title', L._('Zoom out'));

    _qt('.leaflet-control-attribution a[href $= "leafletjs.com"]', 'title', L._('A JavaScript library for interactive maps'));
  }

  /** Only translate the default marker ALT text, for now.
   * @see https://github.com/Leaflet/Leaflet/blob/main/src/layer/marker/Marker.js#L49
   */
  function _localizeMarker (PIN_EL) {
    // const MARKER_PANE = MAP.getPane('markerPane'); // Was: MAP_EL.querySelector('.leaflet-marker-pane');

    // [...MARKER_PANE.children].forEach(PIN_EL => {
    if (PIN_EL.alt === 'Marker') {
      PIN_EL.alt = L._('Marker');
      PIN_EL.title = L._('Marker');
    }
    // });

    // console.debug('localizeMarkers:', MARKER_PANE.children);
  }

  function _localizePopups (MAP) {
    MAP.on('popupopen', ev => {
      ev.popup._closeButton.title = L._('Close popup'); // src/layer/Popup.js#L102
      ev.popup._closeButton.ariaLabel = L._('Close popup');

      console.debug('a11y.popupopen:', ev);
    });
  }

  /**
   * Set a role and role on the map container.
   * @see GH Issue: Leaflet/Leaflet/issues/7193.
   * @see SC 4.1.2: https://w3.org/TR/WCAG21/#name-role-value
   */
  function _fixMapContainer () {
    if (!mapElem.role) mapElem.role = 'region';
    if (!mapElem.ariaLabel) mapElem.ariaLabel = L._('map');
    mapElem.ariaRoleDescription = L._('map');
  }

  /**
   * Fix for non-interactive markers.
   * @see GH Issue: Leaflet/Leaflet/issues/8116
   */
  function _fixMarkers (MAP) {
    let layerIdx = 0;
    MAP.on('layeradd', ev => {
      // const isMarker = ev.layer._icon;
      const markerEl = ev.layer._icon || null;
      const isInteractive = (markerEl && markerEl.classList.contains('leaflet-interactive')) || false;

      if (markerEl && !isInteractive) {
        markerEl.tabIndex = -1;
        markerEl.role = undefined;
        markerEl.classList.add('x-static-marker');
      }
      if (markerEl) {
        _localizeMarker(markerEl);
      }
      console.debug('a11y.layeradd', layerIdx, isInteractive, markerEl, ev);
      layerIdx++;
    });
  }

  /**
   * Move keyboard focus when a popup is opened and closed.
   * @see GH Issue: Leaflet/Leaflet/issues/8115 (also: #8113, #8114)
   * @see SC 2.4.3: https://w3.org/TR/WCAG21/#focus-order
   */
  function _managePopupFocus (MAP) {
    MAP.on('popupopen', (ev) => {
      ev.popup._container.role = 'dialog'; // Not modal!
      ev.popup._container.tabIndex = -1;
      ev.popup._container.focus(); // Was: ev.popup._closeButton.focus();

      const SOURCE = ev.popup._source;
      if (SOURCE) {
        SOURCE._icon.ariaExpanded = true;
      }
    });

    MAP.on('popupclose', (ev) => {
      // Find the marker or element that triggered the popup.
      const SOURCE = ev.popup._source;
      if (SOURCE) {
        SOURCE._icon.focus();
        SOURCE._icon.ariaExpanded = 'false';
      }

      console.debug('a11y.popupclose (F):', SOURCE, ev);
    });
  }

  /** _qt: Find element within map container, and set a property on it, translated with '_()' (i18n plugin).
   */
  function _qt (selector, property, str) {
    const ELEM = mapElem.querySelector(selector);
    console.assert(ELEM, `element.querySelector('${selector}')`);

    if (ELEM) {
      ELEM[property] = str;
    }
  }

  // return your plugin when you are done
  return MyAccessibilityPlugin;
}, window));
